### PR TITLE
New rule to flag special characters in identifiers

### LINKED ISSUE #1746: New rule suggestion: identify and flag special characters in object names
Should these sorts of issues be bugs or enhancements?

We should be able to identify and flag special characters in names when creating objects.

Anti-pattern:
```
CREATE TABLE DBO.ColumnNames
(
    [TrailingSpace ] INT,
    [Greater>Than] INT,
    [Less<Than] INT,
    Number# INT
)
```

### PR DESCRIPTION
**Pull request by jpers36**
Fixes #1746 

### Brief summary of the change made
A new linting rule: flag special characters in identifiers.

If an identifier includes a character other than alphanumeric or underscore, this rule will flag it.
This rule includes configuration settings to control which types of identifiers fall under the rule.
This rule also includes a configuration setting to allow or disallow spaces as part of an identifier.

+L057 python script
+L057 test YAML
+L057 configuration defaults
+allow_space_in_identifier description in config_info.py

### CODE REVIEW

File: src/sqlfluff/rules/L057.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+@document_configuration
+@document_fix_compatible
+class Rule_L057(BaseRule):
+    """Do not use special characters in object names.
+
+    The functionality for this rule is inherited from :obj:`Rule_L010`.
+
+    | **Anti-pattern**
+    | Using special characters within names when creating or aliasing objects.
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:21]:
```suggestion
    | Using special characters within identifier names when creating or aliasing objects.
```

**Review by jpy-git — COMMENTED** [review]
few quick tweaks whilst i look through the rule logic

File: src/sqlfluff/rules/L057.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+)
+from sqlfluff.rules.L014 import identifiers_policy_applicable
+
+
+@document_configuration
+@document_fix_compatible
+class Rule_L057(BaseRule):
+    """Do not use special characters in object names.
+
+    The functionality for this rule is inherited from :obj:`Rule_L010`.
```
[the remark this answers is not in the document]
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git removed, thanks!

File: src/sqlfluff/rules/L057.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+from sqlfluff.core.rules.base import BaseRule, LintResult, RuleContext
+from sqlfluff.core.rules.doc_decorators import (
+    document_configuration,
+    document_fix_compatible,
+)
+from sqlfluff.rules.L014 import identifiers_policy_applicable
+
+
+@document_configuration
+@document_fix_compatible
```
[the remark this answers is not in the document]
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git removed @document_fix_compatible, thanks!

File: src/sqlfluff/rules/L057.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+class Rule_L057(BaseRule):
+    """Do not use special characters in object names.
```
[the remark this answers is not in the document]
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git I standardized on identifier(s).

File: src/sqlfluff/rules/L057.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+"""Implementation of Rule L057."""
+
+from typing import Optional
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, RuleContext
+from sqlfluff.core.rules.doc_decorators import (
+    document_configuration,
+    document_fix_compatible,
```
[the remark this answers is not in the document]
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git removed, thanks!

File: src/sqlfluff/rules/L057.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        CREATE TABLE DBO.ColumnNames
+        (
+            [Internal Space] INT,
+            [Greater>Than] INT,
+            [Less<Than] INT,
+            Number# INT
+        )
+
+    | **Best practice**
+    | Object names should include only alphanumerics and underscores.
```
[the remark this answers is not in the document]
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git I standardized on identifier(s).  Thanks!

File: src/sqlfluff/rules/L057.py (hunk 1 above)
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git I standardized on identifier(s).  Thanks!

File: src/sqlfluff/rules/L057.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                        .replace("_", "")
+                        .replace(" ", "")
+                        .isalnum()
+                    )
+                )
+            )
+        ):
+            return LintResult(anchor=context.segment)
+        else:
+            return None
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:88]: no need for else, just have final return None as a general catch for non-matched segment
```suggestion
        
        return None
```

File: src/sqlfluff/rules/L057.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        if (
+            (
+                context.segment.name == "naked_identifier"
+                and identifiers_policy_applicable(
+                    self.unquoted_identifiers_policy, context.parent_stack  # type: ignore
+                )
+                and not (
+                    context.segment.raw.replace("_", "").isalnum()
+                    or (
+                        self.allow_space_in_identifier  # type: ignore
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:61]: remove type ignore comment

File: src/sqlfluff/rules/L057.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                        .replace(" ", "")
+                        .isalnum()
+                    )
+                )
+            )
+        ) or (
+            (
+                context.segment.name == "quoted_identifier"
+                and identifiers_policy_applicable(
+                    self.quoted_identifiers_policy, context.parent_stack  # type: ignore
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:72]: remove type ignore comment

File: src/sqlfluff/rules/L057.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                )
+                and not (
+                    context.segment.raw[1:-1].replace("_", "").isalnum()
+                    or (
+                        self.allow_space_in_identifier  # type: ignore
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:77]: remove type ignore comment

File: src/sqlfluff/rules/L057.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                        and context.segment.raw[1:-1]
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:78]:
```suggestion
                        and context.segment.raw.strip('"')
```

File: src/sqlfluff/rules/L057.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    config_keywords = [
+        "quoted_identifiers_policy",
+        "unquoted_identifiers_policy",
+        "allow_space_in_identifier",
+    ]
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Do not use special characters in object names."""
+        if (
```
Comment (jpy-git) [inline src/sqlfluff/rules/L057.py:52]: suggest a slight tweak to the structure for readability:
```python
# Check for non-alphanumeric characters in unquoted identifiers:
if <unquoted_identifer_conditions>:
    return LintResult(anchor=context.segment)

# Check for non-alphanumeric characters in quoted identifiers:
if <quoted_identifer_conditions>:
    return LintResult(anchor=context.segment)

return None
```

File: test/fixtures/rules/std_rule_cases/L057.yml [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    )
+
+test_pass_special_chars_create_table_none_quoted_tsql:
+  pass_str:
+    CREATE TABLE DBO.ColumnNames
+    (
+        [No_Internal_Space] INT,
+        [GreaterThan] INT,
+        [LessThan] INT,
+        NumberVal INT
```
Comment (jpy-git) [inline test/fixtures/rules/std_rule_cases/L057.yml:51]: is it possible to have quotes and square brackets or are the two mutually exclusive?
Not a t-sql user myself so curious 😄

File: test/fixtures/rules/std_rule_cases/L057.yml [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    FROM DBO.ColumnNames
+
+test_pass_special_chars_reference:
+  pass_str:
+    SELECT "column with space" as a
+    FROM DBO.ColumnNames
+  configs:
+    rules:
+      L057:
+        quoted_identifiers_policy: aliases
```
Comment (jpy-git) [inline test/fixtures/rules/std_rule_cases/L057.yml:81]: maybe include some test cases with custom quoted_identifiers_policy configs and also some where one of the identifiers policies are set to none to verify they pass as expected

**Review by jpy-git — COMMENTED** [review]
@jpers36 a couple suggestions formatting wise, plus note on typing, and extra unit tests suggestions.
For the most part its looking great! 💪 

(always find it a bit mad that sql allows these non alnum characters)

File: test/fixtures/rules/std_rule_cases/L057.yml (hunk 13 above)
Comment (jpers36) [reply test/fixtures/rules/std_rule_cases/L057.yml:51]: The two are mostly interchangeable.  You wouldn't use both, but one or the other.

File: src/sqlfluff/rules/L057.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            (
+                context.segment.name == "naked_identifier"
+                and identifiers_policy_applicable(
+                    self.unquoted_identifiers_policy, context.parent_stack  # type: ignore
```
[the remark this answers is not in the document]
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git adjusted, thanks!

File: src/sqlfluff/rules/L057.py (hunk 7 above)
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git adjusted, thanks!

File: src/sqlfluff/rules/L057.py (hunk 8 above)
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git adjusted, thanks!

File: src/sqlfluff/rules/L057.py (hunk 9 above)
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git adjusted, thanks!

File: src/sqlfluff/rules/L057.py (hunk 10 above)
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git adjusted, thanks!

File: src/sqlfluff/rules/L057.py (hunk 11 above)
Comment (jpers36) [reply src/sqlfluff/rules/L057.py]: @jpy-git I don't think this change would work.  There are many other quotes used by various dialects, such as TSQL's square brackets.  I believe another dialect uses backticks.  I did confirm when updating L029 a few weeks back that we don't support any dialects with multi-character quotes.

File: test/fixtures/rules/std_rule_cases/L057.yml (hunk 14 above)
Comment (jpers36) [reply test/fixtures/rules/std_rule_cases/L057.yml:158]: @jpy-git test cases have been expanded!

**Review by jpy-git — APPROVED** [review]
Looks good! 🥇

**Comment by jpy-git** [comment]
@barrywhart I don't have write access so can't approve.

I've looked over the code and all feedback has been addressed so can you take a quick look over and if there's nothing else to add then approve and merge 😄

**Review by barrywhart — APPROVED** [review]

**Pull request merged**